### ISSUE TITLE
Use pre-commit and black for formatting

### ISSUE THREAD

**Issue by nicoddemus** [reaktoro/Reaktoro#16]
Want to get opinions if we should start to use [pre-commit](https://pre-commit.com/) for general checkins (like docs, formatting, etc) and [black](https://github.com/ambv/black) for automatic code formatting of Python code.

I've been using both in all my open source projects and have been happy with it. More and more Python projects are also adopting black as their auto formatter.

**Comment by allanleal** [comment]
Hi @nicoddemus . Thanks for the suggestion. The black project seems interesting, and its adoption in Reaktoro will ensure everyone contributing python code does not need to be super concerned about their formatting as this gets corrected automatically. So, we have consistent format across all python codes.

For C++, I know something similar can be achieved with clang. Have you used clang before for formatting C++ codes? We should consider this as well.

I still need a bit more time to read about the pre-commit. I'll get back soon.

**Comment by nicoddemus** [comment]
> For C++, I know something similar can be achieved with clang. Have you used clang before for formatting C++ codes?

AFAIK we never used a formatting tool for C++, but @tadeu will know more details. 👍

**Comment by allanleal** [comment]
Hi @nicoddemus. The pre-commit (and potentially other git hooks) seems like a nice idea. Could you please elaborate a bit more on what we would achieve with pre-commit (a simple/short example)?

**Comment by nicoddemus** [comment]
`pre-commit` allows you to run scripts or tools during each commit to verify syntax, formatting, validation, anything you might want really. This might prevent errors from being introduced in a PR and helps keeping the codebase uniform.

`pre-commit` itself contains a lot of [out-of-the-box hooks](https://github.com/pre-commit/pre-commit-hooks), I think the ones that we could use in Reaktoro are:

* `black` - For Python code formatting
* `blacken-docs` - Ensures code snippets in the docs are also formatted with black
* `end-of-file-fixer` - Makes sure files end in a newline and only a newline.
* `mixed-line-ending` - Replaces or checks mixed line ending (unfortunately this happens more often than it should)
* `trailing-whitespace` - Trims trailing whitespace.

We can also investigate linters for Python and C++ as well, we would need to investigate those.

**Comment by allanleal** [comment]
I see, sounds good to me. Is this something easy to set up for our other repos as well (not necessarily needed now)?

**Comment by nicoddemus** [comment]
Yep, usually is just placing a `.pre-commit-config.yaml` file in the repository and update the developer documentation. 😁

**Comment by allanleal** [comment]
Cool! :)

**Comment by tadeu** [comment]
Just a reminder that, when this is tackled, it would be a nice opportunity to change the encoding of all files to UTF-8 and to normalize all line endings (so that they are all Unix-style LF-only).
